pageserver/src/walredo.rs

```diff
@@ -584,27 +583,168 @@ struct PostgresRedoProcess {
+                    );
+                }
+            } else if err_revents.contains(PollFlags::POLLHUP) {
+                return Err(Error::new(
+                    ErrorKind::BrokenPipe,
+                    "WAL redo process closed its stderr unexpectedly",
+                ));
+            }
+            // If we have some data in stdout, read it to the result buffer.
+            let out_revents = self.poll_fds[0].revents().unwrap();
+            if out_revents & (PollFlags::POLLERR | PollFlags::POLLIN) != PollFlags::empty() {
+                self.page_pos += self.stdout.read(&mut self.page[self.page_pos..])?;
+                if self.page_pos == BLCKSZ as usize {
+                    if let Some(sender) = self.resp_deque.pop_front() {
+                        sender.send(Bytes::copy_from_slice(&self.page)).unwrap();
+                    } else {
+                        return Err(Error::new(
+                            ErrorKind::BrokenPipe,
+                            "Malformed output of walredo process",
+                        ));
+                    }
+                    self.page_pos = 0;
+                }
+            } else if out_revents.contains(PollFlags::POLLHUP) {
+                return Err(Error::new(
+                    ErrorKind::BrokenPipe,
+                    "WAL redo process closed its stdout unexpectedly",
+                ));
+            }
+        }
+        Ok(())
+    }
+
+    #[instrument(skip_all,fields(tenant_id=%self.tenant_id))]
+    fn send(&mut self, sender: Sender<Bytes>, data: Vec<u8>) -> Result<(), Error> {
+        let mut written = 0usize;
+        let data_len = data.len();
+        self.resp_deque.push_back(sender);
+        while written < data_len {
+            let n = loop {
+                match nix::poll::poll(
+                    &mut self.poll_fds[0..3],
+                    self.wal_redo_timeout.as_millis() as i32,
+                ) {
+                    Err(e) if e == nix::errno::Errno::EINTR => continue,
+                    res => break res,
+                }
+            }?;
+
+            if n == 0 {
+                return Err(Error::new(ErrorKind::Other, "WAL redo timed out"));
+            }
+
+            // If we have some messages in stderr, forward them to the log.
+            let err_revents = self.poll_fds[1].revents().unwrap();
+            if err_revents & (PollFlags::POLLERR | PollFlags::POLLIN) != PollFlags::empty() {
+                let mut errbuf: [u8; ERR_BUF_SIZE] = [0; ERR_BUF_SIZE];
+                let n = self.stderr.read(&mut errbuf)?;
+
+                // The message might not be split correctly into lines here. But this is
+                // good enough, the important thing is to get the message to the log.
+                if n > 0 {
+                    panic!(
+                        "wal-redo-postgres: {}",
+                        String::from_utf8_lossy(&errbuf[0..n])
+                    );
```

Comment:
Noted while trying to come up with a benchmark for this (accidential wrong buffertag), that this used to be only an `error!`.

Reply:
Thank you, fixed.